### PR TITLE
Laboratory work #3, Yefimenko Natalya - 23FPL2 

### PR DESCRIPTION
**Pull request by ShrubberyBush**

### CODE REVIEW

File: lab_3_ann_retriever/main.py [hunk 1]
```diff
@@ -50,6 +53,14 @@ def calculate_distance(query_vector: Vector, document_vector: Vector) -> float |
     In case of corrupt input arguments, None is returned.
     """
 
+    if not isinstance(query_vector, list) or not isinstance(document_vector, list):
```
Comment (EIzharikov) [inline lab_3_ann_retriever/main.py:56]: what if vector will be incorrect? for example: `[-1, -100, 0, 5]`

File: lab_3_ann_retriever/main.py [hunk 2]
```diff
@@ -149,6 +181,15 @@ def __init__(self, corpus: list[list[str]]) -> None:
         Args:
             corpus (list[list[str]]): Tokenized documents to vectorize
         """
+        if not (isinstance(corpus, list) and all(isinstance(tokens, list)
```
Comment (EIzharikov) [inline lab_3_ann_retriever/main.py:184]: is it really neede to check types in `__init__`?

File: lab_3_ann_retriever/main.py [hunk 3]
```diff
@@ -158,6 +199,16 @@ def build(self) -> bool:
             bool: True if built successfully, False in other case
         """
 
+        if not self._corpus or not isinstance(self._corpus, list):
+            return False
+        unique_terms = set(token for doc in self._corpus for token in doc)
+        self._vocabulary = sorted(unique_terms)
+        self._token2ind = {word: index for index, word in enumerate(self._vocabulary)}
+        self._idf_values = calculate_idf(self._vocabulary, self._corpus)
+        if self._idf_values is not None and self._vocabulary and self._token2ind:
```
Comment (EIzharikov) [inline lab_3_ann_retriever/main.py:208]: revert if statement; return False if something is wrong and return True at the end if everything is ok!

File: lab_3_ann_retriever/main.py [hunk 4]
```diff
@@ -183,6 +240,12 @@ def vector2tokens(self, vector: Vector) -> list[str] | None:
 
         In case of corrupt input arguments, None is returned.
         """
+        if not isinstance(vector, tuple) or len(vector) != len(self._vocabulary):
+            return None
+        tokens = [self._vocabulary[i] for i, value in enumerate(vector) if value > 0]
+        if not tokens:
+            return None
+        return sorted(tokens)
```
Comment (EIzharikov) [inline lab_3_ann_retriever/main.py:248]: why sorted?

**Review by EIzharikov — COMMENTED** [review]
Nice start, but you have red CI. Fix&answer all my comments and continue your work! You can do it!

File: lab_3_ann_retriever/main.py [hunk 5]
```diff
@@ -171,6 +222,12 @@ def vectorize(self, tokenized_document: list[str]) -> Vector | None:
         In case of corrupt input arguments, None is returned.
         """
 
+        if not (isinstance(tokenized_document, list) and tokenized_document and
+                all(isinstance(token, str) for token in tokenized_document)):
+            return None
+        tf_idf_vector = self._calculate_tf_idf(tokenized_document)
```
[the remark this answers is not in the document]
Comment (ShrubberyBush) [reply lab_3_ann_retriever/main.py]: Done

File: lab_3_ann_retriever/main.py (hunk 4 above)
Comment (ShrubberyBush) [reply lab_3_ann_retriever/main.py:249]: Guess its not necessary

File: lab_3_ann_retriever/main.py (hunk 3 above)
Comment (ShrubberyBush) [reply lab_3_ann_retriever/main.py:210]: Done

File: lab_3_ann_retriever/main.py (hunk 2 above)
Comment (ShrubberyBush) [reply lab_3_ann_retriever/main.py:186]: I guess no...

File: lab_3_ann_retriever/main.py (hunk 1 above)
Comment (ShrubberyBush) [reply lab_3_ann_retriever/main.py:56]: Done

File: lab_3_ann_retriever/main.py [hunk 6]
```diff
@@ -427,6 +574,37 @@ def build(self, vectors: list[Vector]) -> bool:
...
+            return False
+        dimensions = len(vectors[0])
+        another_vec = vectors[:]
+        nodes: list[tuple[list[Vector], int, Node | None, bool]] = [(vectors, 0, None, True)]
+        while nodes:
+            current_vectors, depth, parent, is_left = nodes.pop()
+            if not current_vectors:
+                continue
+            axis = depth % dimensions
+            try:
```
Comment (EIzharikov) [inline lab_3_ann_retriever/main.py:587]: can you do it without these try except cases?

File: lab_3_ann_retriever/main.py [hunk 7]
```diff
@@ -476,6 +658,32 @@ def _find_closest(self, vector: Vector, k: int = 1) -> list[tuple[float, int]] |
         In case of corrupt input arguments, None is returned.
         """
 
+        if not (isinstance(vector, tuple) and isinstance(k, int) and vector
+                and self._root is not None and isinstance(self._root, Node)):
+            return None
+        nearest_neighbors = []
+        nodes: list[tuple[Node, int]] = [(self._root, 0)]
```
Comment (EIzharikov) [inline lab_3_ann_retriever/main.py:665]: remove straight typing

**Review by EIzharikov — COMMENTED** [review]
everything is good, it is quite enough time to make work for 10, you can do it!!! Fix comments

File: lab_3_ann_retriever/main.py (hunk 7 above)
Comment (ShrubberyBush) [reply lab_3_ann_retriever/main.py]: Done

File: lab_3_ann_retriever/main.py (hunk 6 above)
Comment (ShrubberyBush) [reply lab_3_ann_retriever/main.py]: Done

File: lab_3_ann_retriever/main.py [hunk 8]
```diff
@@ -427,6 +574,37 @@ def build(self, vectors: list[Vector]) -> bool:
         In case of corrupt input arguments, False is returned.
         """
 
+        if not vectors or not all(isinstance(vec, tuple) and len(vec) > 0 for vec in vectors):
+            return False
+        dimensions = len(vectors[0])
+        if not all(len(vec) == dimensions for vec in vectors):
+            return False
+        another_vec = vectors[:]
+        nodes: list[tuple[list[Vector], int, Node | None, bool]] = [(vectors, 0, None, True)]
```
Comment (EIzharikov) [inline lab_3_ann_retriever/main.py:583]: remove straight typings

**Review by EIzharikov — COMMENTED** [review]
Good work, fix comment and be ready for Friday

File: lab_3_ann_retriever/main.py (hunk 8 above)
Comment (ShrubberyBush) [reply lab_3_ann_retriever/main.py]: Done

**Pull request closed without merging**